python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -213,14 +190,94 @@ def get_nodes_to_launch(
+                self._request_resources_lift_node_max_launch(
+                    nodes, launching_nodes, max_resources_by_ip,
+                    ensure_min_cluster_size)
+            total_nodes_to_add = self._ensure_min_cluster_size(
+                total_nodes_to_add, ensure_min_cluster_size, node_type_counts,
+                nodes_to_add_based_on_demand)
+        else:
+            request_resources_lift_node_max_launch = {}
         # Limit the number of concurrent launches
         total_nodes_to_add = self._get_concurrent_resource_demand_to_launch(
             total_nodes_to_add, unused_resources_by_ip.keys(), nodes,
-            launching_nodes, nodes_to_add_based_on_requests)
+            launching_nodes, request_resources_lift_node_max_launch)
 
         logger.info("Node requests: {}".format(total_nodes_to_add))
         return total_nodes_to_add
 
+    def _request_resources_lift_node_max_launch(
+            self, nodes: List[NodeID], launching_nodes: Dict[NodeType, int],
+            max_resources_by_ip: Dict[NodeIP, ResourceDict],
+            ensure_min_cluster_size: List[ResourceDict]
+    ) -> Dict[NodeType, int]:
+        """Nodes that can bypass max launch to respect request_resources().
+
+        The maximum number of nodes allowed to launch concurrently to respect
+        request_resources() is the residual of binpacking the desired min
+        cluster size from the max resources of nodes in the heartbeat (active).
+        """
+        # Calculates the residual of binpacking the ensure_min_cluster_size
+        # on the currently running max resources in the heartbeat.
+        current_static_max_resources = copy.copy(
+            list(max_resources_by_ip.values()))
+        resource_requests_to_lift_max_launch, _ = get_bin_pack_residual(
+            current_static_max_resources, ensure_min_cluster_size)
+        # We account here for launching/pending nodes only when used as an
+        # upper bound constraint for max_workers used in get_nodes_for.
+        _, existing_node_type_counts = self.calculate_node_resources(
+            nodes, launching_nodes, {})
+        request_resources_lift_node_max_launch = get_nodes_for(
+            self.node_types, existing_node_type_counts,
+            self.max_workers + 1 - len(nodes),
+            resource_requests_to_lift_max_launch)
+        return request_resources_lift_node_max_launch
+
+    def _ensure_min_cluster_size(
+            self, total_nodes_to_add: Dict[NodeType, int],
+            ensure_min_cluster_size: List[ResourceDict],
+            node_type_counts: Dict[NodeType, int],
+            nodes_to_add_based_on_demand: Dict[NodeType, int]
+    ) -> Dict[NodeType, int]:
+        """Ensure minimum cluster size when using request_resources() API.
+
+        The new nodes to add to ensure minimum cluster size are calculated as
+        follows:
+            1) Get future nodes: pending+launching+running+total_nodes_to_add.
```

Comment:
Got it, so the main issue is we weren't taking into account future nodes launched based on demands?

Reply:
Mainly, yes.